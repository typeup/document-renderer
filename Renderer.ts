import * as dom from "@typeup/dom"
export type renderFunction = (renderer: Renderer, node: dom.Node) => Promise<string>

export class Renderer {
	private constructor(private variables: { [name: string]: string } = {}) {
	}
	getVariable(name: string): string {
		return this.variables[name]
	}
	setVariable(name: string, value: string) {
		this.variables[name] = value
	}
	async render(node: dom.Node | dom.Node[]): Promise<string> {
		let result: string
		if (Array.isArray(node))
			result = (await Promise.all(node.map(n => this.render(n)))).reduce<string[]>((r, e) => Array.isArray(e) ? [...r, ...e] : [...r, e], []).join("")
		else {
			const render = renderers[node.class]
			result = render ? await render(this, node) : ""
		}
		return result
	}
	scope(): Renderer {
		return new Renderer({ ...this.variables })
	}
	static create(): Renderer {
		return new Renderer({})
	}
}
const renderers: { [className: string]: renderFunction } = {}
export function addRenderer(className: string, render: renderFunction) {
	renderers[className] = render
}
