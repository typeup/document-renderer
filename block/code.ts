import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.block.Code): Promise<string> {
	return `<figure>
	<pre><code class="language-${me.language}">
${ me.value }
	</code></pre>
	<figcaption>${ await renderer.render(me.content) }</figcaption>
</figure>
`
}
addRenderer("Block.Code", render)
