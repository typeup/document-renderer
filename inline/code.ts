import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.inline.Code): Promise<string> {
	return `<code>${me.value.replace(/\&/gi, "&amp;").replace(/</gi, "&lt;").replace(/>/gi, "&gt;")}</code>`
}
addRenderer("Inline.Code", render)
