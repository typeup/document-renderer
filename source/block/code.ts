import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.block.Code): Promise<string> {
	return `<figure>
	<pre><code class=${me.language}
	${ me.value }
	</code></pre>
	<figcaption>${ renderer.render(me.content) }</figcaption>
</figure>
`
}
addRenderer("Block.Code", render)
