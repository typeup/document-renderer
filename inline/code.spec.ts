import * as dom from "@typeup/dom"
import { Renderer } from "../Renderer"
import "./code"

jest.useFakeTimers()

const renderer = Renderer.create()
describe("Inline.Code", () => {
	it("render", async () => {
		expect(await renderer.render(new dom.inline.Code("const pi = 3.1415"))).toMatchSnapshot()
	})
	it("html", async () => {
		expect(await renderer.render(new dom.inline.Code("<html></html>"))).toMatchSnapshot()
	})
})
