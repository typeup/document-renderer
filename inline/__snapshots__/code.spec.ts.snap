// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline.Code html 1`] = `"<code>&lt;html&gt;&lt;/html&gt;</code>"`;

exports[`Inline.Code render 1`] = `"<code>const pi = 3.1415</code>"`;
